win32/tests/gballoc_ll_jemalloc_int/gballoc_ll_jemalloc_int.c

```diff
@@ -260,4 +263,131 @@ TEST_FUNCTION(gballoc_ll_print_stats_works)
     gballoc_ll_free(ptr);
 }
 
+static void assert_set_option_works_for_decay(const char* option_name, int64_t decay_ms, uint32_t narenas)
+{
+    char command[32];
+    int snprintf_result;
+    int64_t verify_decay_ms;
+    size_t decay_ms_size = sizeof(decay_ms);
+
+    // Arena 1 does not exist, so je_mallctl should return EFAULT
+    for (uint32_t i = 0; i < narenas; i++)
+    {
+        snprintf_result = snprintf(command, sizeof(command), "arena.%" PRIu32 ".%s_ms", i, option_name);
+        ASSERT_IS_TRUE(snprintf_result > 0 && snprintf_result < sizeof(command), "snprintf failed");
+
+        if (i == 1)
+        {
+            // the arena at index 1 does not exist, so this should fail with EFAULT
+            ASSERT_ARE_EQUAL(int, EFAULT, je_mallctl(command, &verify_decay_ms, &decay_ms_size, NULL, 0));
+        }
+        else
+        {
+            ASSERT_ARE_EQUAL(int, 0, je_mallctl(command, &verify_decay_ms, &decay_ms_size, NULL, 0));
+            ASSERT_ARE_EQUAL(int64_t, decay_ms, verify_decay_ms);
+        }
+    }
+}
+
+static void setup_arenas_for_set_option(uint32_t expected_narenas, void* ptr[])
+{
+    // verify the number of arenas
+    uint32_t narenas;
+    size_t narenas_size = sizeof(narenas);
+    ASSERT_ARE_EQUAL(int, 0, je_mallctl("opt.narenas", &narenas, &narenas_size, NULL, 0));
+    ASSERT_ARE_EQUAL(uint32_t, expected_narenas, narenas);
+
+    ptr[0] = gballoc_ll_malloc(4); // this malloc will happen in the default arena since this thread is linked to arena 0 by default
+    ASSERT_IS_NOT_NULL(ptr[0]);
+
+    // Initialize all the arenas except first(arena at index 1) after the default arena, so that we can verify that the decay is set for all the arenas except arena at index 1
+    uint32_t arena_id;
+    for (uint32_t i = 2; i < narenas; i++)
+    {
+        arena_id = i; // assign thread to arena i
+        ASSERT_ARE_EQUAL(int, 0, je_mallctl("thread.arena", NULL, NULL, &arena_id, sizeof(arena_id)));
+
+        ptr[i] = gballoc_ll_malloc(4); // this malloc will happen in the arena i
+        ASSERT_IS_NOT_NULL(ptr[i]);
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

function should assign the thread back to its original arena

Reply:
Done